### PR TITLE
Prevent infinite loop from cyclic reference, create generic ReferenceResolver

### PR DESCRIPTION
**Pull request by AttilaTheFun**
This PR reworks https://github.com/AttilaTheFun/SwaggerParser/pull/25 to create the generic ReferenceResolver and prevent an infinite loop from cyclic references.

### CODE REVIEW

**Comment by yonaskolb** [comment]
Nice one!
Looking at this though, it might be that this fixes simple cases, but due to the early exit by throwing an error, it might miss circular references that happen further down in the spec. That's why I originally returned that temporary value so the building could continue.

A possible fix may be to navigate all the way down in the heirachy and try and build every property, response and schema on the way. At the moment only top level objects are built

**Comment by AttilaTheFun** [comment]
@yonaskolb Okay, I finally got it working. My final solution was to make Structure's structure optional, and only set it to nil when a cyclic reference is found. I also renamed Structure.structure to Structure.type for better ergonomics. See the screenshot below, and the CrossReferenceTests now pass.
screen shot 2017-07-09 at 5 36 32 pm

**Comment by AttilaTheFun** [comment]
@n8chur Any thoughts on this?

**Comment by yonaskolb** [comment]
Looks great, nice work! If only it were as easy as Swift's use of 'indirect' 😄 

Something that you may want to add is to make the actual optional value of the Structure a private variable and then have a public getter that is non optional and that calls `fatalError` if it's nil.  As long as the builders make sure that the reference is indeed resolved. That way consumers don't have to deal with unwrapping an optional.

Just so you know, I actually ended up rewriting my own Swagger library to better fit my needs. In regards to this problem, I solved it without the builder patter, but had to add a single Reference class (like your Structure), which means the SwaggerSpec isn't structs all the way down... :(

**Comment by n8chur** [comment]
@AttilaTheFun: I agree with @yonaskolb's comment regarding the pubic interface to `Structure`... I would prefer that the stucture's type be non-null as a consumer of the library. I think @yonaskolb's solution or using an intermediary type (where the intermediary structure's type is nullable and the final type is non-null) could address this issue.

File: Sources/ReferenceResolver.swift [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+extension ResponseBuilder: ResolvableType {
+    static var path: String { return "responses" }
+    static var resolver: ReferenceResolver<ResponseBuilder> { return kResponseBuilderReferenceResolver }
+}
+
+
+class ReferenceResolver<T: ResolvableType> {
+    enum ResolverError: Error {
+        case invalidContext
+        case invalidReference
```
Comment (n8chur) [inline Sources/ReferenceResolver.swift]: Is this used?

File: Sources/ReferenceResolver.swift [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        case invalidPath
+        case unsupportedReference
+        case unresolvedReference
+    }
+
+    private class Context<C> {
+        var cachedReferences = [String: C]()
+        var resolvingReferences = [String: Bool]()
```
Comment (n8chur) [inline Sources/ReferenceResolver.swift]: Could you add comments describing how these are used?

File: Sources/ReferenceResolver.swift [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+        if let cached = context.cachedReferences[name] {
+            return Structure(name: name, type: cached)
+        }
+
+        if let resolving = context.resolvingReferences[name], resolving {
+            return Structure(name: name, type: nil)
+        }
+
+        context.resolvingReferences[name] = true
```
Comment (n8chur) [inline Sources/ReferenceResolver.swift]: Could you add a short comment describing how this works / why you set this temporarily? It took me a minute to process why this is set to true here and then nil after building is complete.

File: Sources/ReferenceResolver.swift [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        let resolved = try builder.build(swagger)
```
Comment (n8chur) [inline Sources/ReferenceResolver.swift]: Is an error thrown here if `resolved` is nil?

File: Sources/Swagger.swift [hunk 5]
```diff
@@ -116,19 +116,33 @@ struct SwaggerBuilder: Builder {
...
-            Structure(name: name, structure: try builder.build(swagger))
+            Structure(name: name, type: try builder.build(swagger))
         }
 
         let responses = try self.responses.map { name, builder in
-            Structure(name: name, structure: try builder.build(swagger))
+            Structure(name: name, type: try builder.build(swagger))
         }
 
+        // Clean up resolvers:
```
Comment (n8chur) [inline Sources/Swagger.swift]: Why is this cleanup necessary? Is it purely a memory optimization?

**Comment by n8chur** [comment]
Do we have a test for cyclical references?

**Comment by AttilaTheFun** [comment]
@n8chur I added 'CrossReferenceTests' which has two definitions that reference each other cyclicly. That's what's shown in the screenshot above. 

@yonaskolb Thanks for the feedback re: the optional structure. Yeah I could do the fatalError computed property approach suggested, but I've been using Stencil for templating which internally uses Mirror.descendent which fails for computed properties. I know it's ugly, but maybe the best approach here is an implicitly unwrapped optional? Then it would only crash if you tried to traverse the tree up to that point, but ideally you shouldn't need to go any further than knowing that it references a named structure.

File: Sources/Swagger.swift (hunk 5 above)
Comment (AttilaTheFun) [reply Sources/Swagger.swift]: Yeah, we don't need to persist that information in memory after the resolution has finished.

File: Sources/ReferenceResolver.swift (hunk 4 above)
Comment (AttilaTheFun) [reply Sources/ReferenceResolver.swift]: Resolved is a non-nil T.Building because Builder.build returns a non-optional building or it throws an error.

**Comment by AttilaTheFun** [comment]
@n8chur @yonaskolb  I think I addressed your specific comments, and if the implicitly unwrapped optional approach is amenable to you I can implement that later today.

**Comment by yonaskolb** [comment]
Yeah the implicitly unwrapped optional sounds better.
 
Not having access to getters through Stencil is one of the many reasons I chose to create an intermediate representation in SwagGen. It also takes a lot of the logic out of the templates and into Swift

**Comment by AttilaTheFun** [comment]
@yonaskolb Cool, I can add that later today. Also do you think it's preferable to have the logic in the templates or in Swift? I've been building swift utilities that I can plug in to stencil, but trying to keep as much of the logic in my templates as I can. Also, you mentioned you ended up rewriting your Swagger library to better meet your needs. Are there things I could add to SwaggerParser that would allow you to use it? Feel free to open issues on this library if there are other things you need.

**Comment by yonaskolb** [comment]
@AttilaTheFun I think there's a balance to be struck about how much logic is in code and how much is in the template. All the logic in code makes it easier to write, modify and test, and allows writing logic in one place so it can be used in multiple templates, but makes the templates less flexible. Putting all the logic in the templates makes them really flexible, but they are much harder to write, become fragile, and you end up duplicating things all over the place.
There's also things that are quite easy to do in code but harder in a templating language.

In my opinion, logic that transforms data in a way that can be shared across templates should be in code. I've gone a step further and have a specific formatter for Swift templates that does things like escaping names and types against reserved Swift keys, pre-formatting built in types (eg so a boolean type becomes "Bool"), adding "?" to optional types and other swift related helpers.

I think anything that changes how files are layed out and composed should be in templates, as well as any template specific logic. This is an area to chose to go further in compared to the official Swagger CodeGen. In that project, all the output files are defined in code. In my own SwagGen the output files are defined in a yaml template manifest which is also transformed via stencil filters and custom options

**Comment by AttilaTheFun** [comment]
@yonaskolb That's a really good point. I definitely found myself fighting the templating language a lot, when the code would have been pretty trivial to write in Swift. I'm definitely going to start rewriting that which makes sense as pluggable Stencil filters instead of Stencil macros. E.g. extracting the type name or parameter name from a SwaggerParser Schema.

**Review by yonaskolb — APPROVED** [review]

**Pull request merged**